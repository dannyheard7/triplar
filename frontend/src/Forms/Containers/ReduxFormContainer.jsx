import React from "react";
import {inputParsers} from "../utils/inputParsers.js";
import {connect} from "react-redux";

class ReduxFormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        const data = this.parseFormData(event);

        this.props.dispatch(this.props.action(data));
    }

    parseFormData(event) {
        const form = event.target;
        const data = new FormData(event.target);

        for (let name of data.keys()) {
            const input = form.elements[name];
            const parserName = input.dataset.parse;

            if (parserName) {
                const parser = inputParsers[parserName];
                const parsedValue = parser(data.get(name));
                data.set(name, parsedValue);
            }
        }

        let variables = Object();
        for(let pair of data.entries()) {
           variables[pair[0]] = pair[1];
        }

        return variables;
    }

    render() {
        const props = {
            onSubmit: this.handleSubmit,
            fieldErrors: this.props.fieldErrors,
            nonFieldErrors: this.props.nonFieldErrors

        };

        return React.cloneElement(this.props.children, {...props});
    }
}

ReduxFormContainer.defaultProps = {
    fieldErrors: [],
    nonFieldErrors: []
}

export default connect()(ReduxFormContainer)