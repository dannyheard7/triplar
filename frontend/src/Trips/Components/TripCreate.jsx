import React from "react";

import TripForm from "./TripForm";
import ReduxFormContainer from "../../Forms/Containers/ReduxFormContainer";
import {connect} from "react-redux";
import {createTrip} from "../utils/actions";

export class TripCreate extends React.Component {
    constructor(props) {
        super(props);

        this.onSuccess = this.onSuccess.bind(this);
        this.onClick = this.onClick.bind(this);

        this.state = {
            showTripCreate: false
        };
    }

    onSuccess(data) {
        this.setState({showTripCreate: false});

    };

    onClick() {
        this.setState({showTripCreate: !this.state.showTripCreate});
    }

    render () {
        const errors = this.props.errors ? this.props.errors : [];

        return (
            <div>
                <button className="btn btn-primary" onClick={this.onClick}>New Trip</button>
                {this.state.showTripCreate &&
                    <div className="card">
                        <ReduxFormContainer fieldErrors={errors} action={createTrip}>
                            <TripForm submitLabel="Create" />
                        </ReduxFormContainer>
                    </div>
                }
            </div>
        )
    }
}

const mapStateToProps = (state) =>  ({
    errors: state.errors.CREATE_TRIP
});
export default connect(mapStateToProps)(TripCreate)