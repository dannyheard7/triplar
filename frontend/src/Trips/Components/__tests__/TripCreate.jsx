import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
import {TripCreate} from "../TripCreate";
import ReduxFormContainer from "../../../Forms/Containers/ReduxFormContainer";

const renderer = new ShallowRenderer();

const faker = require('faker');

describe('<TripCreateContainer />', () => {
    const data = {trip: {id: faker.random.number()}};

    const props = {
        dispatch: jest.fn()
    };

    test('renders correctly', () => {
        const result = renderer.render(<TripCreate {...props} />);
        expect(result).toMatchSnapshot();
    });

    // TODO
    test.skip('calls onSuccess after receiving success from <ReduxFormContainer />', (function () {
        const stub = jest.spyOn(TripCreate.prototype, 'onSuccess');
        const container = shallow(<TripCreate  {...props}/>);
        container.setState({showTripCreate: true});
        container.find(ReduxFormContainer).prop('onSuccess')(data);

        expect(stub).toBeCalled();
        expect(props.history.push).toBeCalledWith('/trips/' + data.trip.id);
    }));

    // TODO
    test.skip('onSuccess redirects to trip page', (function () {
        const container = shallow(<TripCreate  {...props}/>);
        container.instance().onSuccess(data);

        expect(props.history.push).toBeCalledWith('/trips/' + data.trip.id);
    }));

    describe('trip create visible', () => {
        let container = null;

        beforeEach(() => {
            container = shallow(<TripCreate  {...props}/>);
            container.setState({showTripCreate: true});
        });

        test('renders <ReduxFormContainer /> component when showTripCreate is true', () => {
            expect(container.find(ReduxFormContainer).length).toEqual(1);
        });

        test('updates state.showTripCreate when create button clicked', () => {
            container.find('.btn').simulate('click');
            expect(container.state('showTripCreate')).toBeFalsy();
        });

        // TODO
        test.skip('does not render <ReduxFormContainer /> after a trip has been created', () => {
            expect(container.find(ReduxFormContainer).length).toEqual(1);
            container.find(ReduxFormContainer).prop('onSuccess')(data);
            container.update();

            expect(container.state('showTripCreate')).toEqual(false);
            expect(container.find(ReduxFormContainer).length).toEqual(0);
        });
    });

});