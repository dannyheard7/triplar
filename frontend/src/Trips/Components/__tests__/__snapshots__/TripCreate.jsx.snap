// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripCreateContainer /> renders correctly 1`] = `
<div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    New Trip
  </button>
</div>
`;
